### DISCUSSION TITLE
How to monkey-patch a package running in Vitest?

### DISCUSSION THREAD

**Discussion by timsuchanek** [vitest-dev/vitest#4500] [Q&A] [answered]
I'm working on https://github.com/baserun-ai/baserun-js, which currently uses `module.require` to require the `openai` package to add automatic instrumentation to it.
That works great in all environments, however we couldn't get it running in Vitest yet.
This is our test file:

```ts
import { test } from 'vitest';
import { baserun } from 'baserun';
baserun.init();

import OpenAI from 'openai';

const openai = new OpenAI();


test('automatically instruments openai chat completion', async () => {
  await openai.chat.completions.create({
    messages: [
      {
        role: 'user',
        content: "What's the weather like in San Francisco, Tokyo, and Paris?",
      },
    ],
    model: 'gpt-3.5-turbo',
  });
});
```

1. What we're currently doing (works in Jest and vanilla NodeJS):
```ts
const openai = module.require('openai')
openai.Completions.prototype.create = ...
```

2. What we've tried instead, this interestingly enough breaks the imports of grpc in our library
```ts
import openai from 'openai'
openai.Completions.prototype.create = ...
```
<img width="851" alt="image" src="https://github.com/vitest-dev/vitest/assets/1094804/469facf3-f715-4fbb-8666-bb82e42bebfc">

3. Just `require` - this option doesn't do anything at all and seems to not affect the imported openai model at all
```ts
const openai = require('openai')
openai.Completions.prototype.create = ...
```

4. await import - doesn't do anything either
```ts
const openai = await import('openai')
openai.Completions.prototype.create = ...
```

Any thoughts on what's the best way to accomplish this "monkey patching" pattern within a Vitest context?
I've tried many different options, but nothing seems to work, so I need some help here.

The same principle applies to OpenTelemetry instrumentation - if anyone got that working with vitest, I'd appreciate a link.

Thanks!

**Comment by sheremet-va** [comment]
Does it happen inside `baserun.init`? From you example, I see that it's called after all imports since static imports are hoisted.

If openai only exposes cjs, then patching requite would work if you call init before importing it

If they expose esm then it should also work unless you patch top level exports, but you also need to do it be before importing

**Reply by timsuchanek** [reply]
Thanks for getting back to me so quickly! Yes it happens inside `baserun.init`. So you're saying patching `require` alone should be enough? That doesn't seem to work for me.
I just created a minimal reproduction. In vanilla Node.js, require works, just not in vitest:
https://github.com/timsuchanek/vitest-reproduction

**Reply by sheremet-va** [reply]
Ok, I made a mistake. import that uses cjs entry still doesn't share a cache with ESM, so it won't work. When you tell me that it works in jest/node.js, you mean that you write code in TS and it's transformed into something like this:

```js
const { baserun } = require('baserun');
baserun.init();

const OpenAI = require('openai');

const openai = new OpenAI();
```

Vitest doesn't do that and we use actual ESM, so playing around with `require` doesn't do anything since it doesn't affect ESM in any way. I would recommend testing with JS files or TS that transforms into ESM, not CJS.

**Reply by sheremet-va** [reply, marked as answer]
This way worked for me: https://stackblitz.com/edit/node-cvpkrb?file=package.json,index.js,openai.js

**Reply by timsuchanek** [reply]
Thank you, that also works for me!
